### ISSUE TITLE
Few question related to LiRA MIA

### ISSUE THREAD

**Issue by Ldoun** [meghdadk/SCRUB#1]
First of all, Thank you for the great work!
While trying out your work, I discovered that your repository does not include the LiRA MIA metric. May I ask for the LiRA MIA code?

Also, I've 2 questions related to LiRA MIA.
Each data example is seen exactly half of the time in the train and half of the time in the test when creating the shadow original models. Then when choosing to create 256 unlearning models shouldn't the number of shadow original models be 512?

If my understanding is correct, in LiRA MIA evaluation the Attacker needs to perform 2,560,000 unlearning to create  2,560,000 shadow unlearned models. And after, the Attacker selects 256 shadow unlearning models per example to perform the LiRA Attack. Is there a reason for creating 10k unlearned models per shadow original model instead of one? (I've read that the strength of the attacker is heavily dependent on the number of shadow original/unlearned models. But isn't the suggested MIA create Gaussian Distribution after the Attacker selects 256 models?)

All replies pointing out my understanding are welcome!
Thank you!!

**Comment by meghdadk** [comment]
Hi @Ldoun 
Please see the response from our co-author Jamie who was most involved with the LiRA experiments:

> While trying out your work, I discovered that your repository does not include the LiRA MIA metric. May I ask for the LiRA MIA code?
The code is basically just run [https://github.com/tensorflow/privacy/tree/master/research/mi_lira_2021](mi_lira_2021) on the unlearning algorithm.

> Also, I've 2 questions related to LiRA MIA.
Not sure if this answers your question but I'll try and explain the process and justification. Ideally, for each example, we want N models where it was unlearned and N models where it wasn't used in training. Let's say the target is N=256. It is pretty easy to create 256 models where each example was out of training (by randomly splitting the training set and only using one half for each model we train). Creating 256 models where each example was unlearned is more challenging. Why? The model is trained on 25K samples, but then unlearning is only performed on a few hundred samples. So this means there will not be enough trained models to create a shadow distribution of size N=256 of unlearning for each example. Instead, we bootstrap this process. We train 512 models, so each example has, on average, been included in half of the training sets. Then, for each of 512 trained models, we perform unlearning on a subset of training examples. We run this unlearning procedure many times on different forget sets for each of the 512 models. Creating >>512 "unlearned models", from which we can then create unlearned shadow distributions for each example.
